libs/pilight/protocols/433.92/eurodomest_switch.c

```diff
@@ -0,0 +1,336 @@
+/*
+	Copyright (C) 2015 woutput
+
+	This file is part of pilight.
+
+	pilight is free software: you can redistribute it and/or modify it under the
+	terms of the GNU General Public License as published by the Free Software
+	Foundation, either version 3 of the License, or (at your option) any later
+	version.
+
+	pilight is distributed in the hope that it will be useful, but WITHOUT ANY
+	WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+	A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with pilight. If not, see	<http://www.gnu.org/licenses/>
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include "../../core/pilight.h"
+#include "../../core/common.h"
+#include "../../core/dso.h"
+#include "../../core/log.h"
+#include "../protocol.h"
+#include "../../core/binary.h"
+#include "../../core/gc.h"
+#include "eurodomest_switch.h"
+
+#define PULSE_MULTIPLIER	3
+#define MIN_PULSE_LENGTH	142
+#define MAX_PULSE_LENGTH	426
+#define AVG_PULSE_LENGTH	284
+#define RAW_LENGTH		50
+#define BINARY_LENGTH		24
+#define LEARN_REPEATS		40
+#define NORMAL_REPEATS		10
+
+static int validate(void) {
+	if(eurodomest_switch->rawlen == RAW_LENGTH) {
+		if(eurodomest_switch->raw[eurodomest_switch->rawlen-1] >= (MIN_PULSE_LENGTH*PULSE_DIV) &&
+		   eurodomest_switch->raw[eurodomest_switch->rawlen-1] <= (MAX_PULSE_LENGTH*PULSE_DIV)) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static void createMessage(int id, int unit, int state, int all, int learn) {
+	eurodomest_switch->message = json_mkobject();
+
+	json_append_member(eurodomest_switch->message, "id", json_mknumber(id, 0));
+
+	if(all == 1) {
+		json_append_member(eurodomest_switch->message, "all", json_mknumber(all, 0));
+	} else {
+		json_append_member(eurodomest_switch->message, "unit", json_mknumber(unit, 0));
+	}
+
+	if(state == 1) {
+		json_append_member(eurodomest_switch->message, "state", json_mkstring("on"));
+	} else {
+		json_append_member(eurodomest_switch->message, "state", json_mkstring("off"));
+	}
+
+	if(learn == 1) {
+		eurodomest_switch->txrpt = LEARN_REPEATS;
+	} else {
+		eurodomest_switch->txrpt = NORMAL_REPEATS;
+	}
+}
+
+static void parseCode(void) {
+	int binary[BINARY_LENGTH], x = 0, i = 0;
+	int len = (int)((double)AVG_PULSE_LENGTH*((double)PULSE_MULTIPLIER/2));
+
+	for(x=0;x<eurodomest_switch->rawlen-2;x+=2) {
+		if(eurodomest_switch->raw[x] > len) {
+			binary[i++] = 0;
+		} else {
+			binary[i++] = 1;
+		}
+	}
+
+	// The last 4 bits contain the unit, state and all information
+
+	int unit = 0;
+	int all = 0;
+	int state = 0;
+
+	if(binary[20] == 0 && binary[21] == 0 && binary[22] == 0 && binary[23] == 0) {
```

Comment:
I understand your suggestion, but in my opinion it's not better than my current implementation, *"because it may be shorter, but it is also harder to understand. The function of this code block is to act on received **bits**, not just "a (decimal) value (without meaning)"."*
(Why) do you think using decimal would be better?